File: NewsAppJava/app/src/main/java/com/avenue/taipt/newsappjava/ui/fragments/ArticleFragment.java
package com.avenue.taipt.newsappjava.ui.fragments;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.WebViewClient;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;

import com.avenue.taipt.newsappjava.databinding.FragmentArticleBinding;
import com.avenue.taipt.newsappjava.models.Article;
import com.avenue.taipt.newsappjava.ui.NewsViewModel;
import com.google.android.material.snackbar.Snackbar;

import java.util.List;

public class ArticleFragment extends Fragment {

    private FragmentArticleBinding binding;
    private NewsViewModel viewModel;
    private Article article;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentArticleBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        viewModel = new ViewModelProvider(requireActivity()).get(NewsViewModel.class);
        article = new Article();
        if (getArguments() != null) {
            article = ArticleFragmentArgs.fromBundle(getArguments()).getArticle();
            binding.webView.setWebViewClient(new WebViewClient());
            binding.webView.loadUrl(article.getUrl());
        }
        binding.fab.setOnClickListener(v -> {
            viewModel.saveArticle(article);
            Snackbar.make(view, "Article saved successfully!", Snackbar.LENGTH_SHORT).show();
        });
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        binding = null;
    }
}

File: NewsApp/app/src/main/java/taipt4/kotlin/newsapp/ui/fragments/ArticleFragment.kt
package taipt4.kotlin.newsapp.ui.fragments

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.webkit.WebViewClient
import androidx.fragment.app.Fragment
import androidx.navigation.fragment.navArgs
import com.google.android.material.snackbar.Snackbar
import taipt4.kotlin.newsapp.databinding.FragmentArticleBinding
import taipt4.kotlin.newsapp.ui.NewsActivity
import taipt4.kotlin.newsapp.ui.NewsViewModel

class ArticleFragment : Fragment() {

    private var _binding: FragmentArticleBinding? = null
    private val binding get() = _binding!!

    lateinit var viewModel: NewsViewModel
    val args: ArticleFragmentArgs by navArgs()

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        Log.d("TaiPT", "ArticleFragment::onCreateView()")
        _binding = FragmentArticleBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        Log.d("TaiPT", "ArticleFragment::onViewCreated()")
        val article = args.article
        binding.webView.apply {
            webViewClient = WebViewClient()
            loadUrl(article.url!!)
        }
        binding.fab.setOnClickListener {
            viewModel.saveArticle(article)
            Snackbar.make(view, "Article saved successfully!", Snackbar.LENGTH_SHORT).show()
        }
    }

    override fun onStart() {
        super.onStart()
        Log.d("TaiPT", "ArticleFragment::onStart()")
        viewModel = (activity as NewsActivity).viewModel
    }

    override fun onDestroyView() {
        super.onDestroyView()
        Log.d("TaiPT", "ArticleFragment::onDestroyView()")
        _binding = null
    }
}